### PR TITLE
Abort Qt builds that do not use a supported deriver

### PR DESCRIPTION
**Pull request by ttuegel**
###### Motivation for this change

This will give package maintainers faster feedback to potential build and runtime problems.

See also: https://github.com/NixOS/nixpkgs/issues/65399#issuecomment-538001109

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Review by bkchr — APPROVED** [review]

**Comment by worldofpeace** [comment]
Things that use `mkDerivationWith` fail to build like `kcc`. How about a check in the qtbase setup hook that wrapQtAppsHook is present?

**Comment by ttuegel** [comment]
> Things that use `mkDerivationWith` fail to build like `kcc`. How about a check in the qtbase setup hook that wrapQtAppsHook is present?

This satisfies things that use `mkDerivationWith`, but it will break things that don't use any Qt deriver. It is not sufficient to only use `wrapQtAppsHook`.

**Review by worldofpeace — APPROVED** [review]
Seems right to me :+1: Aborting seems more friendly TBH.

File: pkgs/development/libraries/qt-5/hooks/qtbase-setup-hook.sh [hunk 1]
```diff
@@ -1,3 +1,13 @@
+if [ -z "$_qt_mkDerivation" ]
+then
+    cat >&2 <<EOF
+Did not detect supported mkDerivation. Build aborted!
+
+Please refer to the Nixpkgs manual section 'Support for specific programming languages and frameworks/Qt'.
```
Comment (samueldr) [inline pkgs/development/libraries/qt-5/hooks/qtbase-setup-hook.sh]: Not a blocker issue. Is there something we can do to improve the message in some way?

Linking to the manual page should be possible, even if not yet online (will be when 19.09 is released). All sections should already have stable IDs, it's a matter of placing it after `https://nixos.org/manual/nixpkgs/`.

Maybe add some keywords to help a forgetful person? Something like "tip: look for qt5.callPackage" or something like that.

**Comment by samueldr** [comment]
This would still be needed.

This would also need to detect using mixed Qt versions.

**Comment by veprbl** [comment]
Would be better to instead make Qt to behave more like a normal nixpkgs library #71089

**Comment by samueldr** [comment]
Sure, though detecting mixed Qt would be needed anyway.

**Comment by ttuegel** [comment]
This is neither necessary nor useful to detect mixed Qt versions.

**Pull request closed without merging**